cpp/node_similarity_module/algorithms/node_similarity.hpp

```diff
@@ -0,0 +1,206 @@
+char const *node1_name = "node1";
+char const *node2_name = "node2";
+char const *similarity = "similarity";
+
+}  // namespace node_similarity_util
+
+
+namespace node_similarity_algs {
+
+
+// Lambda functions for computing concrete similarities
+/*
+Calculates Jaccard similarity between two nodes based on their neighbours.
+ns1 are neighbours of the first node
+ns2 are neighbours of the second node
+*/
+
+double JaccardFunc(const std::set<uint64_t> &ns1, const std::set<uint64_t> &ns2) {
+    std::set<uint64_t> elem_union, elem_intersection;
+    std::set_intersection(ns1.begin(), ns1.end(), ns2.begin(), ns2.end(), std::inserter(elem_intersection, elem_intersection.begin()));
+    std::set_union(ns1.begin(), ns1.end(), ns2.begin(), ns2.end(), std::inserter(elem_union, elem_union.begin()));
+    if (elem_union.size() == 0) {
+        return 0.0;
+    } else {
+        return elem_intersection.size() / (double) elem_union.size();
+    }
+}
+
+/*
+Calculates overlap function between two set of neighbours.
+ns1 are neighbours of the first node.
+ns2 are neighbours of the second node.
+*/
+double OverlapFunc(const std::set<uint64_t> &ns1, const std::set<uint64_t> &ns2) {
+    std::set<uint64_t> elem_intersection;
+    std::set_intersection(ns1.begin(), ns1.end(), ns2.begin(), ns2.end(), std::inserter(elem_intersection, elem_intersection.begin()));
+    int denonominator = std::min(ns1.size(), ns2.size());
+    if (denonominator == 0) {
+        return 0.0;
+    } else {
+        return elem_intersection.size() / (double) denonominator;
+    }
+}
+
+
+
+/*
+Calculates cosine similarity function between two nodes for a given property.
+*/
+double CosineFunc(const mgp::Node &node1, const mgp::Node &node2, const std::string &property) {
+    const auto &prop1_it = node1.GetProperty(property);
+    const auto &prop2_it = node2.GetProperty(property);
+    if (prop1_it.IsNull() || prop2_it.IsNull()) {
+        throw mgp::ValueException("All nodes should have property " + property);
+    }
+    const auto &prop1 = prop1_it.ValueList();
+    const auto &prop2 = prop2_it.ValueList();
+    double similarity = 0.0, node1_sum = 0.0, node2_sum = 0.0;
+    int size1 = prop1.Size(), size2 = prop2.Size();
+    if (size1 != size2) {
+        throw mgp::ValueException("Vectors should be of the same size.");
+    }
+    for (int i = 0; i < size1; ++i) {
+        double val1 = prop1[i].ValueDouble(), val2 = prop2[i].ValueDouble();
+        similarity += val1 * val2;
+        node1_sum += val1 * val1;
+        node2_sum += val2 * val2;
+    }
+    double denominator = sqrt(node1_sum) * sqrt(node2_sum);
+    if (denominator < 1e-9) {
+        return 0.0;
+    } else {
+        return similarity / denominator;
+    }
+}
+
+
+
+/*
+Extract node neighbours.
+*/
+std::set<uint64_t> GetNeighbors(std::unordered_map<uint64_t, std::set<uint64_t>> &neighbors, const mgp::Node &node) {
+    uint64_t node_id = node.Id().AsUint();
+    const auto &result_it = neighbors.find(node_id);
```

Comment:
nice optimization, but it was not clear for me at least, at the beginning what is happening. Maybe you can do the following: change `neighbors` name to `all_node_neighbors` and change order of checks:

```
if (result_it != neighbors.end()) {
    return result_it->second;
        
}
std::set<uint64_t> ns;
const auto &rels = node.OutRelationships();
for (const auto &rel: rels) {
    ns.insert(rel.To().Id().AsUint());
}
neighbors[node_id] = ns;
return ns;
```

neighbors[node_id] = ns; <- Also, this line does a copy.
Maybe you can avoid that, by setting value like this:
neighbors[node_id] = std::set<-->()

neighbors[node_id].insert(..)

And return reference? Or just rename function to `UpdateNeighbors` and don't return anything.

Reply:
I would partially agree. Inserting into map container does create a copy but I think is more performant to do this copy than every time indexing the map with: `neighbors[node_id].insert(..)`. At least this came out from flamegraph stuff with mgp PR.

Reply:
what if you remeber reference to that array:

auto &current_set = neighbors[node_id]

current_set.insert()...

Reply:
https://en.cppreference.com/w/cpp/container/map/operator_at


I think it should work

Reply:
yes I changed it to what you wrote.